fix(client):bug:: Fix button once handler from not working after an error occurs

SBAsai opened the pull request:
This PR will fix the original `button | once` handler from disabling in the event an error occurs. The buttons will now continue to work after an error has appeared while also only calling the attached function once.

SBAsai committed:
fix(client): 🐛 Fix button once from preventing re-click after error

Fixed a bug where a player could not re-click the client-sided buttons if an error is thrown since the event was only allowing one click.

SBAsai committed:
Merge branch 'development' into asai/button-once-fix

SBAsai committed:
fix(client): 🐛 Add button once prevention on more buttons

Added the same bug fix to the create lobby and join lobby buttons on the splash page. Removed the `once` handler from the start game button

SBAsai commented:
Forgot to add the fix to the splash page buttons too

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -17,6 +17,10 @@
   let errorMessage: string = "";
 
+  // variable that will be set true if the corresponding function has no errors thrown
+  // this will then allow the button to be pressed again if there is an error thrown
```
thia comment should use `/** */` because it describes the meaning of a variable

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -17,6 +17,10 @@
 
+  // variable that will be set true if the corresponding function has no errors thrown
+  // this will then allow the button to be pressed again if there is an error thrown
+  let once: boolean = false;
```
I think we could find a more descriptive variable name maybe `buttonActive` or `waiting` or something like that

SBAsai replied:
I'll just name all of them `waiting`

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -91,8 +99,22 @@
+            await createLobbyHandler();
+          }
+        }}
+        disabled={!nameValidation.valid}><Label>Create Lobby</Label></Button
```
this if should be moved into the disabled

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -91,8 +99,22 @@
+            await joinLobbyHandler();
+          }
+        }}
+        disabled={!nameValidation.valid}><Label>Join Lobby</Label></Button
```
this if should be moved into the disabled

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -104,7 +110,13 @@
   {/if}
-  <form on:submit|once={joinLobby} on:submit|preventDefault>
+  <form
+    on:submit|preventDefault={async () => {
+      if (once == false) {
+        await joinLobby();
+      }
+    }}
```
this if should be moved into the disabled if that works for enter

SBAsai replied:
Seems like it works

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -24,6 +24,8 @@
   let codeDirty: boolean = false;
   $: codeValidation = lobbyCodeValidator(code);
 
+  let once: boolean = false;
```
I think we could find a more descriptive variable name maybe `buttonActive` or `waiting` or something like that

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -15,6 +15,10 @@
 
+  // variable that will be set true if the corresponding function has no errors thrown
+  // this will then allow the button to be pressed again if there is an error thrown
+  let once: boolean = false;
```
I think we could find a more descriptive variable name maybe `startingGame` or `waiting` or something like that

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -15,6 +15,10 @@
   export let lobby: Lobby;
   let errorMessage: string = "";
 
+  // variable that will be set true if the corresponding function has no errors thrown
+  // this will then allow the button to be pressed again if there is an error thrown
```
thia comment should use `/** */` because it describes the meaning of a variable

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -77,14 +85,22 @@
   {#if auth.currentUser?.uid === lobby.uids[0]}
     <div class="actions">
-      <Button on:click|once={() => start()}><Label>Start Game</Label></Button>
+      <Button
+        on:click={async () => {
+          if (once == false) {
+            await start();
+          }
+        }}><Label>Start Game</Label></Button
+      >
```
this if should be moved into the disabled

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -77,14 +85,22 @@
   {/if}
   <div class="actions">
     <Button
-      on:click|once={async () => {
+      on:click={async () => {
+        if(once == false) {
         await leave();
         goto("/");
+        }
       }}><Label>Leave Lobby</Label></Button
     >
```
this if should be moved into the disabled

SBAsai committed:
refactor(client): ♻️ Move button if statements into `disabled`

Moved the if statements into `disabled=`. Changed `once` variable name to `waiting`. Updated comments

SBAsai committed:
style(client): 🎨 Format with prettier

Formatted with prettier

SBAsai committed:
Merge branch 'development' into asai/button-once-fix

gchamb reviewed (dismissed):
Looks good to me. The user could hypothetically remove the disabled property and then spam but idk if there's a way to check serverside to rate limit their requests

nstringham commented:
> Looks good to me. The user could hypothetically remove the disabled property and then spam but idk if there's a way to check serverside to rate limit their requests

I feel like server side rate limiting is outside the scope of this card

gchamb commented:
> I feel like server side rate limiting is outside the scope of this card

I would agree just throwing the idea out there

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -74,15 +84,21 @@
     <div class="actions">
-      <Button on:click|once={() => start()}><Label>Start Game</Label></Button>
+      <Button
+        on:click={async () => {
+          await start();
+        }}
```
this can be `() => start()` like it was before this PR

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -15,6 +15,12 @@
   export let lobby: Lobby;
   let errorMessage: string = "";
 
+  /**
+  variable that will be set true if the corresponding function has no errors thrown
+  this will then allow the button to be pressed again if there is an error thrown
+  */
```
each line should start with a `* `

 ```ts
  /**
   * variable that will be set true if the corresponding function has no errors thrown
   * this will then allow the button to be pressed again if there is an error thrown
   */
 ```

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -91,8 +101,18 @@
-      <Button on:click|once={createLobbyHandler} disabled={!nameValidation.valid}><Label>Create Lobby</Label></Button>
-      <Button on:click|once={joinLobbyHandler} disabled={!nameValidation.valid}><Label>Join Lobby</Label></Button>
+      <Button
+        on:click={async () => {
+          await createLobbyHandler();
+        }}
```
this can just be `on:click={createLobbyHandler}` like it was before

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -91,8 +101,18 @@
+        disabled={!nameValidation.valid || waiting}><Label>Create Lobby</Label></Button
+      >
+      <Button
+        on:click={async () => {
+          await joinLobbyHandler();
+        }}
```
this can just be `on:click={joinLobbyHandler}` like it was before

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -17,6 +17,12 @@
 
   let errorMessage: string = "";
 
+  /**
+  variable that will be set true if the corresponding function has no errors thrown
+  this will then allow the button to be pressed again if there is an error thrown
+  */
```
see other comment about doc comment formating

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -24,6 +24,12 @@
   let codeDirty: boolean = false;
   $: codeValidation = lobbyCodeValidator(code);
 
+  /**
+  variable that will be set true if the corresponding function has no errors thrown
+  this will then allow the button to be pressed again if there is an error thrown
+  */
```
see other comment about doc comment formating

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -54,18 +60,22 @@
       // if the lobby doesn't exist then error is thrown
       errorMessage = err instanceof Error ? err.message : String(err);
       if (errorMessage == "You are already in the lobby!") {
+        waiting = true;
```
what does this do?

SBAsai replied:
If you are already in the lobby, the button should be disabled. Otherwise other errors being thrown should keep `waiting = false`

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -104,7 +114,14 @@
+        await joinLobby();
+      }
+    }}
+    disabled={waiting}
```
this should probobly be on the submit button instead of the form

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -104,7 +114,14 @@
-  <form on:submit|once={joinLobby} on:submit|preventDefault>
+  <form
+    on:submit|preventDefault={async () => {
+      if (waiting == false) {
+        await joinLobby();
+      }
```
Do you need this `if` if you alread have the disabled?

SBAsai committed:
style(client): 🎨 Change button onclick function calls

Changed function calls on various buttons

SBAsai committed:
style(client): 🎨 Fix comment formatting

Fixed how block comments are formatted

SBAsai committed:
fix(client): 🐛 Fix disable button on join page

Moved the button disable from join lobby form to the button

SBAsai committed:
Merge branch 'development' into asai/button-once-fix

SBAsai committed:
Merge branch 'development' into asai/button-once-fix

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -85,17 +95,18 @@
   {#if $user?.uid === lobby.uids[0]}
     <div class="actions">
-      <Button on:click|once={() => start()} disabled={lobby.players.length < minPlayers}
+      <Button on:click|once={() => start()} disabled={lobby.players.length < minPlayers || waiting}
```
why is this still `once`

nstringham commented on client/src/routes/+page.svelte:
```diff
@@ -91,8 +101,10 @@
-      <Button on:click|once={joinLobbyHandler} disabled={!nameValidation.valid}><Label>Join Lobby</Label></Button>
+      <Button on:click={createLobbyHandler} disabled={!nameValidation.valid || waiting}
+        ><Label>Create Lobby</Label></Button
+      >
```
I would prefer if this was formatted like
```html
<Button on:click={createLobbyHandler} disabled={!nameValidation.valid || waiting}>
  <Label>Create Lobby</Label>
</Button>
```

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -104,7 +115,11 @@
+  <form
+    on:submit|preventDefault={async () => {
+        await joinLobby();
+      }}
```
this can just be `joinLobby` instead of `async () => { await joinLobby(); }`

nstringham reviewed (dismissed):
Seams good just a few tiny changes

SBAsai committed:
refactor(client): ♻️ Remove `once` from start game function call

Removed the `| once` call from the start game function call

SBAsai committed:
style(client): 🎨 Change button formats

Changed button formatting and indentation

SBAsai committed:
refactor(client): ♻️ Fix button function calls

Fixed how the functions are called inside of the `onclick`s

SBAsai committed:
Merge branch 'development' into asai/button-once-fix

nstringham commented on client/src/routes/join/+page.svelte:
```diff
@@ -116,10 +116,8 @@
-        await joinLobby();
-      }}
-  >
+    on:submit|preventDefault={joinLobby}
```
can this be all on one line

SBAsai committed:
style(client): 🎨 Change form-matting

Changed the formatting on a form element
